File: 백준/Gold/1520. 내리막 길/내리막 길.js
const fs = require('fs');
const input = fs
  .readFileSync(process.platform === 'linux' ? '/dev/stdin' : 'index.txt')
  .toString()
  .trim()
  .split('\n');

const [N, M] = input[0].split(' ').map(Number);
const map = input.slice(1).map(line => line.split(' ').map(Number));

const dp = Array.from({ length: N }, () => Array(M).fill(-1));

function dfs(x, y) {
    if (x === N - 1 && y === M - 1) return 1; 
    if (dp[x][y] !== -1) return dp[x][y]; 

    dp[x][y] = 0; 

    for (const [dx, dy] of [[0,1],[1,0],[-1,0],[0,-1]]) {
        const nx = x + dx;
        const ny = y + dy;

        if (
            0 <= nx && nx < N &&
            0 <= ny && ny < M &&
            map[nx][ny] < map[x][y]
        ) {
            dp[x][y] += dfs(nx, ny);
        }
    }

    return dp[x][y];
}

console.log(dfs(0, 0));

File: 백준/Gold/1520. 내리막 길/내리막 길.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N, M = map(int, input().split())

A = [list(map(int, input().split())) for _ in range(N)]
dp = [[-1] * M for _ in range(N)]

dx = [0,0,1,-1]
dy = [1,-1,0,0]

def dfs(x, y) :
    if x == N-1 and y == M -1 :
        return 1
    
    if dp[x][y] != -1 :
        return dp[x][y]
    
    dp[x][y] = 0

    for i in range(4) :
        nx, ny = x + dx[i], y + dy[i]
        if 0 <= nx < N and 0 <= ny < M :
            if A[x][y] > A[nx][ny] :
                dp[x][y] += dfs(nx,ny)
    return dp[x][y]

print(dfs(0,0))